### DISCUSSION TITLE
filter requires spaces between multiple search terms to work

### DISCUSSION THREAD

**Discussion by spiderbit** [davatorium/rofi#1824] [Q&A] [answered]
Lets say I have a list and want to search a match out of multiple words: let's say:

```
 foo bar baz
 foo baz baaz
 foo bar b123
 foo bzz baaz
```

let's say I want to find the second line with:
`ba baa`

Is there no way that I don't need the space in between and can't just search
`babaa`

Emacs smex / fuzzy search does that with buffers. Is that not possible or is there a trick to configure rofi to do that? As example with the windowswitcher.

**Comment by DaveDavenport** [comment, marked as answer]
Try `-matching fuzzy`

**Comment by spiderbit** [comment]
Yes that does the trick thx, did search for fuzzy but somehow had -fuzzy in the options instead of the -matching fuzzy.

**Comment by spiderbit** [comment]
@DaveDavenport there is still a problem, because the window titel are extremly long it finds multiple matches even for long substrings. Can I limit the characters per Colums or in this case the title so instead of {w}  {t} something like {w}:10   {t}:20 or so.

I tried with themes but even if it cuts of parts of the string with ... at the end it still uses the stuff after that for filtering.

**Reply by DaveDavenport** [reply]
No, it searches everything, not what is visible. It would be weird if matching would change based on font size/window width etc.
Fuzzy is a 'fuzzy' matcher.

**Reply by spiderbit** [reply]
Yes doing that with the theme would be weird I agree but I would just want to only have a "max width" setting where it only shows and searches that. Because fuzzy works bad if you have a book length of words for each entry, because then all entries matches every letter combination, fuzzy works better with smaller length entries < 30 or so.

That's why I used in emacs exwm that in the exwm-update-title-hook:
```emacs-lisp
(defun sb/build-buffer-name ()
     "Shortens to long buffer names"
     (if (> (length exwm-title) 30)
	 (let* ((pos (string-match "[\\|-] .\\\{3,25\\\}$" exwm-title )))
	   (concat (substring exwm-title 0 25)
		   (if pos (format " [%s]"
				   (car (last (split-string exwm-title))) (+ 2 pos )))))
       exwm-title))

(add-hook 'exwm-update-title-hook
	     (lambda ()
	       (when (or (not exwm-instance-name)
                 ...
		 (exwm-workspace-rename-buffer (sb/build-buffer-name)))))
```
Mostly that is relevant if you use a no-tab browser add-on, like >100 character long titles for most browser windows. As example for this website browser window it displays:
`[ program name | the whole title of the post - then the name of the repository | discussion 1824 | ... more stuff at least the browser name ]`

I guess it's less about showing only parts of the name but actively renaming them, I guess I could try to do that in stumpwm.

Just looked into it, the api of stumpwm gives me hooks that might or might not work `*new-window-hook*` but stumpwms api don't exposes me a rename function at least not easily or obviously. But it does expose a `*window-format*` varible which does what I want but only for the internal window-list functions:

> Note, a prefix number can be used to crop the argument to a specified size. For instance, ‘%20t’ crops the window’s title to 20 characters. 

But using stumpwm's internal windowlist is also no real solution because A it's inconsistent with the looks of other rofi prompts, it's limited to theming so far that I even failed to change the font yet.

**Reply by DaveDavenport** [reply]
This feels like a very niche request, and I want to avoid adding 1024 different options. 
I think this can be resolved by making a small user (mode) script that lists all the window titles (truncated) and allows you to switch to them. I don't know  stumpwm at all, but maybe it has a small client (like i3/herbstluftwm) that can do most of the heavy lifting.

**Reply by spiderbit** [reply]
Well yes I tried it, just wanted to look how far I can come with rofi options because I did not succeed till now with using stumpish to get the windowlist , but I figured it out, probably don't need the stumpish cli tool in the end, but that works:
```common-lisp
(setq *window-menu*
     (loop for key being the hash-keys of *xwin-to-window*
        using (hash-value value)
           collect (cons (let ((title (window-title value))
			       (length 30))
			   (if (> (length title) length)
			       (subseq title 0 length)
			       title))
			 (format nil
				 "exec stumpish pull-window-by-number ~A"
				 (window-number value)))))

(defcommand draw-window-menu () ()
  ;; (rofi:menu *window-menu*)
  ;; or with rofi arguments
  (rofi:menu *window-menu* "-i -matching fuzzy -show-icons -font 'Dejavu Sans Mono 20' -theme gruvbox-dark-hard"))

(define-key *root-map* (kbd "Menu") "draw-window-menu")
```

That's at least a first working prototype (depends on stumpwm-rofi modul).

**Reply by spiderbit** [reply]
I still find it slightly strange that I have to find a theme that has the desired length/width, why can't I use as example solarized theme but just 1/3 as broad of a box. Why is there room for like 500 characters. I can define the amount of lines with -l but not the width? -n? or it should maybe orientate on the longest entry in the list and not make it >3 times as big.

**Reply by DaveDavenport** [reply]
I  am unsure what you are asking. I think you want to know why rofi does not auto size its width based on the content?

Loading of the content happens asynchronous from displaying. When rofi loads it does not know how wide to make the window.
If the visible bit of the content is loaded, should it resize to that?  this means  the window width can change when you scroll through the options? (as new ones can be added) .

Rendering an entry (to determine its width)  is also an expensive operation, if you feed rofi with 10000 entries, you don't want to have to render all just to size the window. 


Btw you can easily set the window width on the commandline:

```bash
rofi -show combi -theme-str 'window { width: 20em;}'`
````
or

```bash
rofi -show combi -theme+window+width 20em
```

**Reply by spiderbit** [reply]
ok thanks that should do the trick

**Comment by spiderbit** [comment]
Ok just if somebody tries my posted code here, window-menu can't be a defined as a one time variable with setq, it must be a function otherwise it never updates the window list:

```common-lisp
(defun stump-windows () 
...
)

(rofi:menu (stump-windows) "-show run -i -matching fuzzy -theme gruvbox-dark -font 'Dejavu Sans Mono 22' -theme+window+width 20em "))
```
